### PR TITLE
Setting up Karate Test for GraphQl

### LINKED ISSUE #29: Set up Karate Testing Framework for GraphQL API - Users Module
Set up automated testing for our Hasura GraphQL API using Karate framework, focusing initially on the users and authentication functionality.

Follow this [tutorial](https://hasura.io/blog/testing-hasura-graphql-apis-with-karate)
 
## Problem
We need to:
- Set up Karate testing environment
- Test user-related GraphQL operations
- Verify Firebase auth integration
- Test users
- Generate test reports

## Solution

A possible solution looks like:

### Project Structure
```plaintext
/tests
├── docker-compose.test.yml
├── karate-config.js
├── logback-test.xml
├── karate/
│   ├── features/
│   │   ├── auth/
│   │   │   ├── login.feature
│   │   │   └── permissions.feature
│   │   └── users/
│   │       ├── create.feature
│   │       ├── query.feature
│   │       └── wallets.feature
│   └── helpers/
│       └── schemas.js
└── results/
```

### Example Test Feature
```gherkin
# karate/features/users/query.feature
Feature: User Queries

Background:
  * url HASURA_ENDPOINT
  * def query = read('queries/list-users.graphql')

Scenario: Query Current User
  Given header Authorization = 'Bearer ' + firebase_token
  And header x-hasura-role = 'user'
  And request { 
    query: 'query GetCurrentUser { users_by_pk(id: "' + user_id + '") { id email last_seen } }'
  }
  When method post
  Then status 200
  And match response.data.users_by_pk.id == user_id
  And match response.errors == '#notpresent'

Scenario: Query User Wallets
  Given header Authorization = 'Bearer ' + firebase_token
  And def query = read('queries/get-user-wallets.graphql')
  And def variables = { userId: '#(user_id)' }
  And request { query: '#(query)', variables: '#(variables)' }
  When method post
  Then status 200
  And match response.data.user_wallets[0] contains { is_primary: true }
  And match response.errors == '#notpresent'
```

### Example GraphQL Queries
```graphql
# karate/features/users/queries/get-user-wallets.graphql
query GetUserWallets($userId: String!) {
  user_wallets(
    where: {user_id: {_eq: $userId}}
    order_by: {is_primary: desc}
  ) {
    id
    wallet_address
    chain_type
    is_primary
    created_at
  }
}
```

### User Authentication
- [ ] Test Firebase token validation
- [ ] Verify JWT claims
- [ ] Test invalid tokens
- [ ] Check token expiration

### User Queries
- [ ] Get current user
- [ ] List user roles
- [ ] Query user wallets
- [ ] Test permission boundaries

### User Roles
- [ ] Verify tenant role assignments
- [ ] Verify landlord role assignments
- [ ] Test role-based access control
- [ ] Test role transitions

### User Wallets
- [ ] Query wallet information
- [ ] Verify primary wallet logic
- [ ] Test wallet address uniqueness
- [ ] Test chain type validation

### Error Scenarios
- [ ] Missing authentication
- [ ] Invalid user IDs
- [ ] Unauthorized access
- [ ] Invalid wallet addresses

### Docker Configuration
```yaml
version: '3.8'
services:
  karate:
    image: karate-test
    build:
      context: .
      dockerfile: Dockerfile.test
    volumes:
      - ./karate:/karate
      - ./results:/results
    environment:
      HASURA_ENDPOINT: http://graphql-engine:8080/v1/graphql
      HASURA_ADMIN_SECRET: ${HASURA_ADMIN_SECRET}
      FIREBASE_API_KEY: ${FIREBASE_API_KEY}
    depends_on:
      - graphql-engine
```

### Karate Config with Firebase Support
```javascript
// karate-config.js
function fn() {
  var env = karate.env || 'dev';
  var config = {
    env: env,
    HASURA_ENDPOINT: java.lang.System.getenv('HASURA_ENDPOINT'),
    HASURA_ADMIN_SECRET: java.lang.System.getenv('HASURA_ADMIN_SECRET'),
    FIREBASE_API_KEY: java.lang.System.getenv('FIREBASE_API_KEY')
  };
  
  // Setup Firebase test tokens
  if (env == 'dev') {
    // Add test user tokens
    config.tenant_token = 'test-tenant-token';
    config.landlord_token = 'test-landlord-token';
  }
  
  return config;
}
```

## Success Criteria
- [ ] User query tests passing
- [ ] Role verification working
- [ ] Wallet tests implemented
- [ ] Error scenarios covered
- [ ] Test reports generated

## Prerequisites
- [ ] Test users created
- [ ] Test wallets configured

### PR DESCRIPTION
**Pull request by olisaagbafor**
This PR is being superseded by #56 to align with project commit guidelines.

# Pull Request for SafeTrust - Add Automated API Testing Framework

❗ **Pull Request Information**

This PR implements comprehensive automated testing for the SafeTrust GraphQL API using the Karate framework. It establishes a robust testing infrastructure focusing on user authentication, permissions, and wallet functionality.

## 🌀 Summary of Changes

- **Test Infrastructure Setup**:
  - Added Karate testing framework configuration
  - Created Docker test environment
  - Implemented test reporting system

- **Test Suite Implementation**:
  - Authentication tests (Firebase token validation, JWT claims)
  - User management tests (CRUD operations, permissions)
  - Wallet functionality tests (creation, validation)
  - Error scenario coverage

- **Project Structure**:
  - Organized test files in feature-based hierarchy
  - Added reusable test helpers and schemas
  - Implemented environment-specific configurations

### Evidence Before Solution

- No automated testing infrastructure
- Manual verification required for API changes
- Inconsistent test coverage

### Evidence After Solution

- Automated test execution via Docker
- Comprehensive test coverage for:
  - User authentication flows
  - RBAC permissions
  - Wallet operations
  - Error scenarios
- Generated test reports showing coverage metrics

## 📂 Related Issue

This pull request will **close #29** - Implement automated API testing infrastructure

---

🎉 Thank you for reviewing this PR!

### CODE REVIEW

**Comment by diegoTech14** [comment]
Hi @olisaagbafor 

1. Could you change the dependencies from package.json in the tests directory to the main package.json?
2. We would like to see the setup working in a video. Could you make us a video, please?

For the video you can use loom -> https://www.loom.com/looms/videos

**Comment by diegoTech14** [comment]
@olisaagbafor Hey man, PR looks good, just would like to see a video please 🚀

**Review by diegoTech14 — APPROVED** [review]

**Comment by olisaagbafor** [comment]
@diegoTech14 The video link is [here](https://www.loom.com/share/78c1130e709747ce8f02bc6004a9c358?sid=e0d43388-43e3-40eb-90d3-e9a169a0ea60)

**Comment by diegoTech14** [comment]
> @diegoTech14 The video link is [here](https://www.loom.com/share/78c1130e709747ce8f02bc6004a9c358?sid=e0d43388-43e3-40eb-90d3-e9a169a0ea60)

Will check ASAP, thanks!

File: tests/docker-compose.test.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
+services:
+  postgres:
```
Comment (rvalenciano) [inline tests/docker-compose.test.yml:2]: This looks great, however lets move this to the root folder as well, also lets rename it as `docker-compose-test.yml`

File: tests/docker-compose.test.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      timeout: 5s
+      retries: 5
+      start_period: 30s
+    networks:
+      - safetrust-network
+    volumes:
+      - postgres_data:/var/lib/postgresql/data
+    command: ["postgres", "-c", "logging_collector=on", "-c", "log_statement=all"]
+
+  graphql-engine:
```
Comment (rvalenciano) [inline tests/docker-compose.test.yml:22]: Lets rename this to `graphql-engine-test`

File: tests/docker-compose.test.yml (hunk 1 above)
Comment (rvalenciano) [inline tests/docker-compose.test.yml:2]: Lets also rename this to `postgres_test`

File: tests/src/test/java/karate/TestRunner.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package karate;
```
Comment (rvalenciano) [inline tests/src/test/java/karate/TestRunner.java:1]: Lets avoid using Java for this, can we set this up with NodeJS?

**Review by rvalenciano — CHANGES_REQUESTED** [review]
@olisaagbafor this is looking great, some mino comments;

Also, can we add some instructions on the README on how to run the tests?

Best

**Comment by diegoTech14** [comment]
@olisaagbafor Hey man, any updates?

**Comment by olisaagbafor** [comment]
Yeah, am on it, Just had some buds to clear up...

**Review by sotoJ24 — APPROVED** [review]

**Comment by zleypner** [comment]
Hello @olisaagbafor , 
May please have your Telegram or contact me at https://t.me/anwar_zlcr ?
I am having an error while running the Karate Tests.

**Comment by zleypner** [comment]
Hello @olisaagbafor! As per our conversation in telegram. Do you have any updates regarding the errors for the tests?

**Comment by olisaagbafor** [comment]
@zleypner Yeah I do will update soon

**Comment by olisaagbafor** [comment]
@zleypner the errors should be gone now, and I have updated the README for the instruction on how to run the test

**Comment by olisaagbafor** [comment]
Please let me know if there is anything else...

**Comment by zleypner** [comment]
@olisaagbafor Thank you for the updates! Checking

**Comment by zleypner** [comment]
Hello @olisaagbafor!
 After running Im getting this command:
docker compose -f docker-compose-test.yml up --build --abort-on-container-exit
I am getting this error
Attaching to graphql-engine-test-1, karate-1, postgres_test-1
Gracefully stopping... (press Ctrl+C again to force)
Error response from daemon: driver failed programming external connectivity on endpoint backend-postgres_test-1 (70ebc50c36c63af19d5d1241effdb232e5b3552201e6aa805e5f94496d856768): Bind for 0.0.0.0:5432 failed: port is already allocated

**Comment by zleypner** [comment]
We are almost there @olisaagbafor , thank you! Would yo be so kind to please update the Loom video?

**Comment by olisaagbafor** [comment]
Ok will do that shortly @zleypner

**Comment by olisaagbafor** [comment]
@zleypner the video link is [here](https://www.loom.com/share/0584af1616404c80b2e0bf2f6feef270?sid=03bf9bc3-05c2-4d2a-8c41-cc0c6cd5bc50)

**Comment by zleypner** [comment]
Thanks @olisaagbafor for this outstanding work! The implementation of the Karate testing framework is extremely well-structured and thorough. I particularly appreciate:

Your responsiveness to feedback and willingness to iterate on improvements
The clear documentation and testing instructions you've added to the README
Your professional approach to addressing all review comments and technical issues

Your attention to detail and commitment to quality really shows through in this PR. This testing infrastructure will be invaluable for maintaining code quality as we scale. Great job on delivering such a robust solution! 🎯 👏

**Comment by olisaagbafor** [comment]
@zleypner thanks alot for your complements ❤️, I truly appreciate 🙏

**Comment by zleypner** [comment]
@olisaagbafor There is one detail pending mentioned in the Contributors guideline. Please update the atomic commits based on the git guidelines. 
image
image

https://github.com/safetrustcr/Frontend/issues/35

**Pull request closed without merging**